### ISSUE TITLE
axis_index crashes XLA inside shmap

### ISSUE THREAD

**Issue by shoyer** [google/jax#15723]
### Description

```python
import jax
import jax.numpy as jnp
from jax.experimental import shard_map
import chex
import functools

chex.set_n_cpu_devices(2)
mesh = jax.sharding.Mesh(jax.devices(), axis_names=('x',))
spec = jax.sharding.PartitionSpec('x')
sharding = jax.sharding.NamedSharding(mesh, spec)

@functools.partial(
    shard_map.shard_map, mesh=mesh, in_specs=(spec,), out_specs=spec,
)
def f(x):
  i = jax.lax.axis_index('x')
  return x + i

x = jax.device_put(jnp.arange(8), sharding)
f(x)
```

Results in the following error:
```
XlaRuntimeError                           Traceback (most recent call last)
[<ipython-input-5-56445fcbbc31>](https://colab.corp.google.com/drive/1KM4xCpYMN4ThadFn5GCm3bR659WPW6hM?resourcekey=0-fmtnoiaRmrEYg-PYM35r_g#) in <module>()
     18 
     19 x = jax.device_put(jnp.arange(8), sharding)
---> 20 f(x)

[<ipython-input-5-56445fcbbc31>](https://colab.corp.google.com/drive/1KM4xCpYMN4ThadFn5GCm3bR659WPW6hM?resourcekey=0-fmtnoiaRmrEYg-PYM35r_g#) in f(x)
     14 )
     15 def f(x):
---> 16   i = jax.lax.axis_index('x')
     17   return x + i
     18 

XlaRuntimeError: INTERNAL: RET_CHECK failure (third_party/tensorflow/compiler/xla/hlo/ir/hlo_instruction.cc:2829) ShapeUtil::CompatibleIgnoringFpPrecision(shape(), new_producer->shape()) s32[2]{0} is not compatible with s32[1]{0}
```

### What jax/jaxlib version are you using?

jax 0.4.9

### Which accelerator(s) are you using?

CPU or TPU

**Comment by shoyer** [comment]
Similar to https://github.com/google/jax/issues/15385, this can be fixed by decorating with `@jax.jit`

**Comment by mattjj** [comment]
This might be a CPU-specific issue, since we have CI testing for `shmap`-of-`axis_index` that is [skipped on CPU](https://github.com/google/jax/blob/8a5a326110bb89f21fc26ebb341087cd0ccdfcc6/tests/shard_map_test.py#L414).

But you've seen this same error on TPU?

**Comment by mattjj** [comment]
This doesn't repro on CPU anymore. I think it's fixed! (There's still #15780 which seems to persist with `cond`, though...)
